### PR TITLE
Merge urlValidation into integration

### PR DESCRIPTION
**Pull request by xaterz**
What's new in this branch:
* Fixed Issue #66, #69, #71 for all pages with GET inputs. Any URL with invalid or missing inputs should now return 404 page instead of crashing the app.
* Implemented new functions in model.py (to be used for all classes):
    * validate_input - validate that the required input values in GET/POST are present and valid
    * get_all_ay_sems - get a list of all the ay sems in the system
* Changed "Modules Taken Prior to Others" to use only GET method
* Implemented POST validation for some pages (edit, restore, internship and non-overlap). Please help me check that all the functions in these pages are still working as intended.

UPDATE:
- Extracted the functions out from model.py into two new files: **database.py** (for database functions), and **helper.py** (for non-database functions). model.py now acts as a facade for these 2 files.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -65,5 +65,10 @@
 '''
 APP = web.application(URLS, globals())
 
+def notfound():
+    return web.notfound(str(RENDER.notfound(message="The page you specified with the URL does not exist.")))
+
+APP.notfound = notfound
+
 if __name__ == '__main__':
```
Comment (nlzz22) [inline app.py]: please pylint this file.

File: components/handlers/module_restore.py [hunk 2]
```diff
@@ -18,16 +18,18 @@ def POST(self, *test_data):
...
+        module_code = input_data.code
+        restore_type = input_data.restoreType
 
-        if restore_type == "quota":
-            ay_sem = data.targetAySem
-            quota = data.quota
+        if restore_type.lower() == "quota":
+            model.validate_input(input_data, ["aysem"])
+            ay_sem = input_data.aysem
+
```
Comment (nlzz22) [inline components/handlers/module_restore.py]: trailing whitespace

File: components/handlers/module_specific_size.py [hunk 3]
```diff
@@ -67,24 +53,20 @@ def GET(self):
         if not session.validate_session():
             raise web.seeother('/login')
 
-        input_data = web.input()
-        if len(input_data) == 0 or input_data.sem is None or input_data.lowerClassSize is None \
-            or input_data.higherClassSize is None:
-            # Loading the page without sufficient data (first time load page)
-            return RENDER.moduleSpecificSize(self.all_ay_sems, None, None, None, None)
-        else:
+        input_data = model.validate_input(web.input(), ["aysem"], aysem_specific=False, attr_required=False)
```
Comment (nlzz22) [inline components/handlers/module_specific_size.py]: line too long

File: components/handlers/module_taken_prior_to_others.py [hunk 4]
```diff
@@ -16,80 +16,42 @@ class TakePriorTo(object):
         This class contains the implementations of the GET
```
Comment (nlzz22) [inline components/handlers/module_taken_prior_to_others.py]: Please pylint this file

File: components/handlers/modules_taken_prior_to_internship.py [hunk 5]
```diff
@@ -16,16 +16,8 @@ class TakePriorInternship(object):
         This class contains the implementations of the GET
```
Comment (nlzz22) [inline components/handlers/modules_taken_prior_to_internship.py]: please pylint this file

File: components/handlers/non_overlapping_modules.py [hunk 6]
```diff
@@ -15,15 +15,9 @@ class NonOverlappingModules(object):
         This class contains the implementations of the GET
```
Comment (nlzz22) [inline components/handlers/non_overlapping_modules.py]: please pylint this file

File: components/model.py [hunk 7]
```diff
@@ -3,10 +3,12 @@
     Handles queries to the database
```
Comment (nlzz22) [inline components/model.py]: please pylint this file

File: test/test_edit_module_fixture.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 '''
```
Comment (nlzz22) [inline test/test_edit_module_fixture.py]: please pylint this file

File: test/test_modules_before_internship.py [hunk 9]
```diff
@@ -25,8 +25,8 @@ class TestCode(object):
 
 
     URL_MODS_BEFORE_INTERNSHIP = '/moduleTakenPriorToInternship'
-    URL_MODS_BEFORE_INTERNSHIP_16_17_SEM2 = '/moduleTakenPriorToInternship?sem=AY%2016/17%20Sem%202'
-    URL_MODS_BEFORE_INTERNSHIP_INVALID = '/moduleTakenPriorToInternship?sem=AY%2016/18%20Sem%202'
+    URL_MODS_BEFORE_INTERNSHIP_16_17_SEM2 = '/moduleTakenPriorToInternship?aysem=AY%2016/17%20Sem%202'
```
Comment (nlzz22) [inline test/test_modules_before_internship.py]: line too long

File: components/model.py [hunk 10]
```diff
@@ -16,6 +18,9 @@
 INDEX_SECOND_ELEM = 1
 LENGTH_EMPTY = 0
 
+# Currently, the system only has data for AY 16/17 and AY 17/18 
+NUMBER_OF_AY_SEMS_IN_SYSTEM = 2
```
Comment (helloqx) [inline components/model.py]: number_of_AY more accurate?

File: components/model.py [hunk 11]
```diff
@@ -18,8 +18,9 @@
 INDEX_SECOND_ELEM = 1
```
Comment (nlzz22) [inline components/model.py]: please fix this linting issue: Line 7: standard import "import hashlib" comes before "import web" (wrong-import-order)

**Comment by xaterz** [comment]
Update:
- I extracted the functions out from model.py into two new files: **database.py** (for database functions), and helper.py (for non-database functions). model.py now acts as a facade for these 2 files.
- When doing nosetests I realised that some test files are calling DB_CURSOR directly. I don't know who wrote them but please don't do that in the future. I have implemented new database functions for those test files to use.
- I also made some changes for edit module to make it even more secure.

Please review the new changes and check that nothing is broken.

**Comment by nlzz22** [comment]
image
I just found out that the following url in the picture above does not result in a 404 page. i.e. as long as the last aysem parameter is valid, it does not matter what invalid strings or items user enters in between.

**Comment by xaterz** [comment]
Regarding the issue with duplicate parameters in the URL, the last value indicated for that parameter will always overwrite the previous ones. This is expected behaviour from web.py.

Since it does not pose any security concern or possibility of crashing the app, this issue can be safely ignored.

**Comment by xaterz** [comment]
Added more comments in validate_input to explain the usage of the optional arguments

**Pull request merged**